Kotlin: ArcUndeclaredThrowableException since 1.13.x

### Describe the bug

Hello,
since Quarkus 1.13.x an ArcUndeclaredThrowableException is thrown when the method has no annotation @Throws(Exception::class) and an exception is thrown by called method.

My code is written in Kotlin and it works in Quarkus 1.12.x but not in 1.13.x.

Exception thrown :
```
2021-11-19 09:41:38,139 | ERROR | executor-thread-0 | [org.acme.quarkus.arcexception.config.exception.RestThrowableMapper:30] - Throwable unhandled (ArcUndeclaredThrowableException): Error invoking subclass method: io.quarkus.arc.ArcUndeclaredThrowableException: Error invoking subclass method
        at ...
Caused by: org.acme.quarkus.arcexception.partner.exception.PartnerException: Error when retrieving partner status: : RESTEASY004655: Unable to invoke request: java.net.UnknownHostException: Hôte inconnu (unknownhost)
        at org.acme.quarkus.arcexception.partner.PartnerApiService.fetchAppStatus(PartnerApiService.kt:19)
        at ...
Caused by: javax.ws.rs.ProcessingException: RESTEASY004655: Unable to invoke request: java.net.UnknownHostException: Hôte inconnu (unknownhost)
        at ...
Caused by: java.net.UnknownHostException: Hôte inconnu (unknownhost)
        at java.base/java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
        at java.base/java.net.InetAddress$PlatformNameService.lookupAllHostAddr(InetAddress.java:929)
        at java.base/java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1519)
        at java.base/java.net.InetAddress$NameServiceAddresses.get(InetAddress.java:848)
        at java.base/java.net.InetAddress.getAllByName0(InetAddress.java:1509)
        at java.base/java.net.InetAddress.getAllByName(InetAddress.java:1368)
        at java.base/java.net.InetAddress.getAllByName(InetAddress.java:1302)
        at org.apache.http.impl.conn.SystemDefaultDnsResolver.resolve(SystemDefaultDnsResolver.java:45)
        at org.apache.http.in.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:112)
        at ...
```

Thanks 

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

You can find a reproducer at https://github.com/Itoriem/quarkus-kotlin-ArcUndeclaredThrowableException

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

"11.0.8" 2020-07-14 LTS Java(TM) SE Runtime Environment 18.9 (build 11.0.8+10-LTS) Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.8+10-LTS, mixed mode)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

1.13

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.3

### Additional information

Discussed here: https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/Kotlin.3A.20ArcUndeclaredThrowableException.20since.201.2E13.2Ex

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/SubclassGenerator.java
@@ -63,6 +63,7 @@ public class SubclassGenerator extends AbstractGenerator {
     private static final DotName JAVA_LANG_THROWABLE = DotNames.create(Throwable.class.getName());
     private static final DotName JAVA_LANG_EXCEPTION = DotNames.create(Exception.class.getName());
     private static final DotName JAVA_LANG_RUNTIME_EXCEPTION = DotNames.create(RuntimeException.class.getName());
+    private static final DotName KOTLIN_METADATA_ANNOTATION = DotNames.create("kotlin.Metadata");
 
     static final String SUBCLASS_SUFFIX = "_Subclass";
     static final String DESTROY_METHOD_NAME = "arc$destroy";
@@ -740,6 +741,7 @@ public class SubclassGenerator extends AbstractGenerator {
         // catch exceptions declared on the original method
         boolean addCatchRuntimeException = true;
         boolean addCatchException = true;
+        boolean isKotlin = method.declaringClass().classAnnotation(KOTLIN_METADATA_ANNOTATION) != null;
         Set<DotName> declaredExceptions = new LinkedHashSet<>(method.exceptions().size());
         for (Type declaredException : method.exceptions()) {
             declaredExceptions.add(declaredException.name());
@@ -763,7 +765,8 @@ public class SubclassGenerator extends AbstractGenerator {
             catchRuntimeException.throwException(catchRuntimeException.getCaughtException());
         }
         // now catch the rest (Exception e) if not already caught
-        if (addCatchException) {
+        // this catch is _not_ included for Kotlin methods because Kotlin has not checked exceptions contract
+        if (addCatchException && !isKotlin) {
             CatchBlockCreator catchOtherExceptions = tryCatch.addCatch(Exception.class);
             // and wrap them in a new RuntimeException(e)
             catchOtherExceptions.throwException(ArcUndeclaredThrowableException.class, "Error invoking subclass method",

--- a/independent-projects/arc/tests/pom.xml
+++ b/independent-projects/arc/tests/pom.xml
@@ -45,6 +45,13 @@
             <artifactId>assertj-core</artifactId>
         </dependency>
 
+        <dependency>
+            <groupId>org.jetbrains.kotlin</groupId>
+            <artifactId>kotlin-stdlib</artifactId>
+            <version>1.5.31</version>
+            <scope>test</scope>
+        </dependency>
+
     </dependencies>
 
     <build>

--- a/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/KotlinInterceptorTest.java
+++ b/independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/interceptors/KotlinInterceptorTest.java
@@ -1,0 +1,71 @@
+package io.quarkus.arc.test.interceptors;
+
+import static java.lang.annotation.ElementType.METHOD;
+import static java.lang.annotation.ElementType.TYPE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.test.ArcTestContainer;
+import java.io.IOException;
+import java.lang.annotation.Documented;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import javax.annotation.Priority;
+import javax.enterprise.context.ApplicationScoped;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.Interceptor;
+import javax.interceptor.InterceptorBinding;
+import javax.interceptor.InvocationContext;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class KotlinInterceptorTest {
+
+    @RegisterExtension
+    public ArcTestContainer container = new ArcTestContainer(Converter.class, ToUpperCaseConverter.class,
+            FailingInterceptor.class, AlwaysFail.class);
+
+    @Test
+    public void testInterceptionThrowsUnwrapped() {
+        ToUpperCaseConverter converter = Arc.container().instance(ToUpperCaseConverter.class).get();
+        assertThrows(IOException.class, () -> converter.convert("holA!"));
+    }
+
+    interface Converter<T> {
+
+        T convert(T value);
+
+    }
+
+    @AlwaysFail
+    @ApplicationScoped
+    @kotlin.Metadata
+    static class ToUpperCaseConverter implements Converter<String> {
+
+        @Override
+        public String convert(String value) {
+            return value.toUpperCase();
+        }
+    }
+
+    @Target({ TYPE, METHOD })
+    @Retention(RUNTIME)
+    @Documented
+    @InterceptorBinding
+    public @interface AlwaysFail {
+    }
+
+    @AlwaysFail
+    @Priority(1)
+    @Interceptor
+    static class FailingInterceptor {
+
+        @AroundInvoke
+        Object fail(InvocationContext ctx) throws Exception {
+            throw new IOException();
+        }
+
+    }
+
+}

